File: Libraries/ImageEncoder/native/csharp-app/source/ImageEncoderUtil.cs
using System;
using System.Collections.Generic;
using Interpreter.Structs;

namespace Interpreter.Libraries.ImageEncoder
{
    internal static class ImageEncoderUtil
    {
        /*
            Format:
                1 - PNG
                2 - JPEG
            
            Return status codes:
                0 - OK
        */
        public static int Encode(object imageObj, int format, List<Value> output, Value[] bytesAsValues)
        {
            System.Drawing.Bitmap image = (System.Drawing.Bitmap)imageObj;

            byte[] bytes = null;
            System.IO.MemoryStream ms = new System.IO.MemoryStream();
            switch (format)
            {
                case 1:
                    image.Save(ms, System.Drawing.Imaging.ImageFormat.Png);
                    bytes = ms.ToArray();
                    break;

                case 2:
                    image.Save(ms, System.Drawing.Imaging.ImageFormat.Jpeg);
                    bytes = ms.ToArray();
                    break;

                default:
                    throw new Exception(); // Library code sent an invalid format flag.
            }
            int length = bytes.Length;
            int i = 0;
            while (i < length)
            {
                output.Add(bytesAsValues[bytes[i++]]);
            }

            return 0;
        }
    }
}

File: Libraries/ImageEncoder/native/java-app/source/ImageEncoderUtil.java
package org.crayonlang.libraries.imageencoder;

import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import javax.imageio.ImageIO;
import org.crayonlang.interpreter.structs.*;

public class ImageEncoderUtil {

	public static int encode(Object imageObj, int formatEnum, ArrayList<Value> output, Value[] bytesAsValues) {
		BufferedImage image = (BufferedImage) imageObj;
		String format;
		switch (formatEnum) {
			case 1: format = "png"; break;
			case 2:
				format = "jpg";
				
				// Re-encode the JPEG to remove any possible alpha channel, which gives corrupted colors.
				int width = image.getWidth();
				int height = image.getHeight();
				BufferedImage newImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
				int[] pixels = image.getRGB(0, 0, width, height, null, 0, width);
				newImage.setRGB(0, 0, width, height, pixels, 0, width);
				image = newImage;
				break;
			default: throw new RuntimeException();
		}
		
		byte[] bytes;
		try {
			ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
			ImageIO.write(image, format, outputStream);
			bytes = outputStream.toByteArray();
		} catch (IOException ioe) {
			return 1;
		}
		
		int i = 0;
		int length = bytes.length;
		int uByte;
		while (i < length) {
			uByte = bytes[i++];
			if (uByte < 0) uByte += 256;
			output.add(bytesAsValues[uByte]);
		}
		
		return 0;
	}
}
